### PR TITLE
Provide configuration variable to access deactivated plugins/groups

### PR DESCRIPTION
**Pull request by marckhouzam**
### What this PR does / why we need it

To test their plugins before activating them in the central repo, publishers can now set the environment variable `TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=true`

This variable tells the CLI to include any deactivated plugins and groups as if they were active.

### Which issue(s) this PR fixes

Fixes # N/A

### Describe testing done for PR

I confirm that `make e2e-context-tmc-test` passes locally.

Then I did the following:
```
$ make plugin-build-and-publish-packages
$ make inventory-init
$ make inventory-plugin-add
$ make inventory-plugin-group-add PLUGIN_GROUP_NAME_VERSION=test:v0.0.1

# Configure the CLI to use the local test repository
$ tz config set env.TANZU_CLI_PRE_RELEASE_REPO_IMAGE localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest
$ tz config set env.TANZU_CLI_PLUGIN_DISCOVERY_IMAGE_SIGNATURE_VERIFICATION_SKIP_LIST localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest

# Make sure we can see the plugins and plugin group
$ tz plugin search
[i] Reading plugin inventory for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest", this will take a few seconds.
[!] Skipping the plugins discovery image signature verification for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"

  NAME     DESCRIPTION             TARGET  LATEST
  builder  Build Tanzu components  global  v0.90.0-alpha.0
  test     Test the CLI            global  v0.90.0-alpha.0
$ tz plugin group search
  GROUP
  vmware-tzcli/test:v0.0.1

# Deactivate plugins
$ make inventory-plugin-deactivate
$ tz plugin search
[i] Reading plugin inventory for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest", this will take a few seconds.
[!] Skipping the plugins discovery image signature verification for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"

  NAME  DESCRIPTION  TARGET  LATEST

# => test that we can see the hidden plugin with the new env var
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=1 tz plugin search
  NAME     DESCRIPTION             TARGET  LATEST
  builder  Build Tanzu components  global  v0.90.0-alpha.0
  test     Test the CLI            global  v0.90.0-alpha.0

# => check that using 'false' will be respected
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=false tz plugin search
  NAME  DESCRIPTION  TARGET  LATEST

# Test installing an active group that contains hidden plugins
$ tz plugin install --group vmware-tzcli/test:v0.0.1
[!] unable to install plugin 'builder': unable to find plugin 'builder' for target 'global'
[!] unable to install plugin 'test': unable to find plugin 'test' for target 'global'
[x] : could not install 2 plugin(s) from group 'vmware-tzcli/test:v0.0.1'

# => Try installing using the group and enabling deactivated plugins
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=true tz plugin install --group vmware-tzcli/test:v0.0.1
[i] Installing plugin 'builder:v0.90.0-alpha.0' with target 'global'
[i] Installing plugin 'test:v0.90.0-alpha.0' with target 'global'
[ok] successfully installed all plugins from group 'vmware-tzcli/test:v0.0.1'

# Hide the group
$ make inventory-plugin-group-deactivate PLUGIN_GROUP_NAME_VERSION=test:v0.0.1
$ tz plugin group search
[i] Reading plugin inventory for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest", this will take a few seconds.
[!] Skipping the plugins discovery image signature verification for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"

  GROUP

# => test that we can see the hidden plugin with the new env var
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=1 tz plugin group search
  GROUP
  vmware-tzcli/test:v0.0.1

# => check that using '0' will be respected
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=0 tz plugin group search
  GROUP

# => Try installing using the hidden group
$ tz plugin install --group vmware-tzcli/test:v0.0.1
[x] : could not find group 'vmware-tzcli/test:v0.0.1'

# => Try installing using the group and enabling deactivated plugins and groups
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=true tz plugin install --group vmware-tzcli/test:v0.0.1
[i] Installing plugin 'builder:v0.90.0-alpha.0' with target 'global'
[i] Installing plugin 'test:v0.90.0-alpha.0' with target 'global'
[ok] successfully installed all plugins from group 'vmware-tzcli/test:v0.0.1'
```
Test publishing a group containing hidden plugins:
```
$ make inventory-plugin-group-add PLUGIN_GROUP_NAME_VERSION=withhidden:v0.0.1
/Users/kmarc/git/tanzu-cli/bin/builder inventory plugin-group add \
		--repository localhost:5001/test/v1/tanzu-cli/plugins \
		--plugin-inventory-image-tag latest \
		--publisher tzcli \
		--vendor vmware \
		--manifest /Users/kmarc/git/tanzu-cli/artifacts/plugins/plugin_group_manifest.yaml \
		--name withhidden:v0.0.1 \

2023-04-20T16:07:26-04:00 [i] pulling plugin inventory database from: "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"
2023-04-20T16:07:28-04:00 [i] updating plugin inventory database with plugin group entry
Error: error while inserting plugin group 'withhidden:v0.0.1': specified plugin 'name:builder', 'target:global', 'version:v0.90.0-alpha.0' is not present in the database
2023-04-20T16:07:28-04:00 [x] : error while inserting plugin group 'withhidden:v0.0.1': specified plugin 'name:builder', 'target:global', 'version:v0.90.0-alpha.0' is not present in the database
make: *** [inventory-plugin-group-add] Error 1
$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=true make inventory-plugin-group-add PLUGIN_GROUP_NAME_VERSION=withhidden:v0.0.1
/Users/kmarc/git/tanzu-cli/bin/builder inventory plugin-group add \
		--repository localhost:5001/test/v1/tanzu-cli/plugins \
		--plugin-inventory-image-tag latest \
		--publisher tzcli \
		--vendor vmware \
		--manifest /Users/kmarc/git/tanzu-cli/artifacts/plugins/plugin_group_manifest.yaml \
		--name withhidden:v0.0.1 \

2023-04-20T16:07:49-04:00 [i] pulling plugin inventory database from: "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"
2023-04-20T16:07:52-04:00 [i] updating plugin inventory database with plugin group entry
2023-04-20T16:07:52-04:00 [i] publishing plugin inventory database
2023-04-20T16:07:54-04:00 [i] successfully published plugin inventory database at: "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"
$ tz plugin group search
[i] Reading plugin inventory for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest", this will take a few seconds.
[!] Skipping the plugins discovery image signature verification for "localhost:5001/test/v1/tanzu-cli/plugins/plugin-inventory:latest"

  GROUP
  vmware-tzcli/withhidden:v0.0.1

$ tz plugin install --group vmware-tzcli/withhidden:v0.0.1
[!] unable to install plugin 'builder': unable to find plugin 'builder' for target 'global'
[!] unable to install plugin 'test': unable to find plugin 'test' for target 'global'
[x] : could not install 2 plugin(s) from group 'vmware-tzcli/withhidden:v0.0.1'

$ TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=true  tz plugin install --group vmware-tzcli/withhidden:v0.0.1
[i] Installing plugin 'builder:v0.90.0-alpha.0' with target 'global'
[i] Installing plugin 'test:v0.90.0-alpha.0' with target 'global'
[ok] successfully installed all plugins from group 'vmware-tzcli/withhidden:v0.0.1'
```

### Release note

```release-note
The `TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=1` variable can now be used to ask the CLI to include deactivated plugins and groups.
```

#### Special notes for your reviewer

**Note 1**
I did something kind of special to simplify the unit tests.

It was becoming increasingly difficult to test `listPluginsFromInventory()` and `listGroupsFromInventory()` now that we had to consider hidden plugins.  The stubbed inventory logic would have to be pretty advanced to be able to properly filter out hidden plugins when needed.

The top commit takes a different approach to testing the `oci_dbbacked.go`file.  We use the fact that we don't need to test the logic implemented by the inventory stub but only the fact that the stub should be called with the correct plugin filter. Instead of returning a filtered list of plugins, the stub returns the plugin/group filter by using the error object; the test can then verify that the filter is appropriate to provide the type of plugins/groups that is expected.

I feel this type of testing is sufficient since there are other tests for the real plugin inventory code which is its own package.

**Note 2**
If a publisher creates an active plugin group containing some deactivated plugins to prepare the group with the idea of activating some plugins later on, the user will see an error that some plugins (the deactivated ones) are missing.  This is not a good UX but since it is a rare case which is not obvious to fix, I will open an separate issue if this PR gets merged as is.

### CODE REVIEW

**Review by anujc25 — COMMENTED** [review]
LGTM. I will wait for other PRs to get merged and this PR to be rebased before I approve this one.

**Comment by marckhouzam** [comment]
Ok, this one is ready.
It is modified slightly to require the use of `TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY=true` when publishing a plugin group that wants to contain deactivated plugins.

**Comment by marckhouzam** [comment]
I've updated the README.md file of the builder plugin to mention the use of the `TANZU_CLI_INCLUDE_DEACTIVATED_PLUGINS_TEST_ONLY` variable.

**Review by anujc25 — APPROVED** [review]
Great work. LGTM. Thanks!

**Pull request merged**
